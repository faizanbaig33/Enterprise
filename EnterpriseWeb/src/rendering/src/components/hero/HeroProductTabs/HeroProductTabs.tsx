// Global
import { useEffect, useState, useRef } from 'react';
import { Feature } from 'src/.generated/Feature.EnterpriseWeb.model';
import { useTheme } from 'lib/context/ThemeContext';
import { Placeholder, withDatasourceCheck, Text } from '@sitecore-jss/sitecore-jss-nextjs';
// Components
import { Component } from 'src/helpers/Component';
import { HeroProductTabsTheme } from './HeroProductTabs.theme';
import { Tabs, TabList, TabPanel, resetIdCounter } from 'react-tabs';
import classNames from 'classnames';
import { SliderWrapper } from 'src/helpers/SliderWrapper';
import { useCurrentScreenType } from 'lib/utils/get-screen-type';

export type HeroProductTabsProps =
  Feature.EnterpriseWeb.Components.Hero.HeroProductTabs.HeroProductTabs;
const HeroProductTabs = (props: HeroProductTabsProps) => {
  const { fields } = props;
  resetIdCounter();
  const slidesToShow = 2;
  const initialSlide = 0;
  const numberOfTabs = props.fields?.children.length;

  const componentRef = useRef<HTMLDivElement>(null);
  const getHashAnchor = (): number => {
    let hashedTabIndex = -1;
    if (typeof window !== 'undefined' && window.location.hash !== '') {
      const hash = window.location.hash.replace('#', '');
      hashedTabIndex = props.fields?.children.findIndex((child: HeroProductTabsProps) => {
        return child.id === hash || child.fields?.contentId?.value === hash;
      });
    }
    return hashedTabIndex;
  };

  let defaultTab = 0;
  const hashedTab = getHashAnchor();
  if (hashedTab > -1) {
    defaultTab = hashedTab;
  } else if (!!props.fields?.defaultActiveTab) {
    const defaultTabIndex = props.fields?.children.findIndex((child: HeroProductTabsProps) => {
      return child.id === props.fields?.defaultActiveTab?.id;
    });
    defaultTab = defaultTabIndex > 0 ? defaultTabIndex : defaultTab;
  }

  const { screenType } = useCurrentScreenType();

  useEffect(() => {
    function handleWindowResize() {
      // set specific height of slick-track so tabs can fill height based on other tabs content
      document.querySelectorAll<HTMLElement>('.heroproducttabs .slick-track').forEach((track) => {
        track.style.height = track.offsetHeight.toString() + 'px';
      });
    }

    window.addEventListener('resize', handleWindowResize);

    if (getHashAnchor() > -1 && componentRef.current) {
      componentRef.current.scrollIntoView();
    }
    return () => {
      window.removeEventListener('resize', handleWindowResize);
    };
  }, []);

  const sliderSettings = {
    infinite: false,
    speed: 500,
    initialSlide: initialSlide,
    slidesToShow: numberOfTabs > slidesToShow ? slidesToShow : 2,
    slidesToScroll: 1,
    dots: false,
    arrows: false,
    variableWidth: true,
    className: 'heroproducttabs',
  };

  const TabHeading = ({ onClick, children, index }: HeroProductTabsProps) => {
    return (
      <div
        className={classNames(
          themeData.classes.tabs.commonTabHeadingClass,
          index === tabIndex
            ? themeData.classes.tabs.selectedTabHeadingClass
            : themeData.classes.tabs.unselectedTabHeadingClass
        )}
        onClick={onClick}
      >
        {children}
      </div>
    );
  };
  TabHeading.tabsRole = 'Tab';

  const [tabIndex, setTabIndex] = useState(defaultTab);

  const { themeData } = useTheme(HeroProductTabsTheme(screenType ?? 'sm', numberOfTabs));

  const isStickyTabonPage = fields.sticky.value;
  // Sticky Tabs
  useEffect(() => {
    if (isStickyTabonPage) {
      window.addEventListener('scroll', isStickyTabs);
    }
    return () => {
      if (isStickyTabonPage) {
        window.removeEventListener('scroll', isStickyTabs);
      }
    };
  });
  const isStickyTabs = () => {
    const heroproducttabsClass = document.querySelector('.heroproducttabs-wrapper') as HTMLElement;
    const utilityNav = document.querySelector('.utility-nav') as HTMLElement;
    const mainNav = document.querySelector('.main-nav') as HTMLElement;
    const mobileHeader = document.querySelector('.mobile-header') as HTMLElement;
    const mainDiv = document.querySelector('main') as HTMLElement;
    const scrollTop = window.scrollY;
    if (scrollTop >= 10) {
      heroproducttabsClass.classList.add('fixed');
      heroproducttabsClass.classList.add('top-0');
      utilityNav.classList.add('hidden');
      mainNav.classList.remove('fixed');
      mainNav.classList.add('static');
      mobileHeader.classList.remove('fixed');
      mobileHeader.classList.add('static');
      if (screenType === 'md' || screenType === 'sm') {
        mainDiv.style.paddingTop = '0';
      }
    } else {
      heroproducttabsClass.classList.remove('fixed');
      heroproducttabsClass.classList.remove('top-0');
      utilityNav.classList.remove('hidden');
      mainNav.classList.remove('static');
      mobileHeader.classList.remove('static');
      mobileHeader.classList.add('fixed');
      if (screenType === 'md' || screenType === 'sm') {
        mainDiv.style.paddingTop = '55px';
      }
    }
  };

  if (!fields) return null;

  return (
    <Component padding={'0'} variant="full" dataComponent="hero/heroproducttabs" {...props}>
      <div className=" col-span-12" ref={componentRef}>
        <Tabs selectedIndex={tabIndex} onSelect={(index: number) => setTabIndex(index)}>
          <div className="heroproducttabs-wrapper z-[70] w-full border-b border-b-gray bg-white">
            <div className="flex w-full flex-col items-center justify-between ml:flex-row ml:px-m lg:mx-auto lg:max-w-screen-lg">
              <div className="w-full ml:w-[50%]">
                <Text
                  tag={'h1'}
                  className={themeData.classes.productHeadlineClass}
                  field={fields.headline}
                />
              </div>
              <div className="ml-auto w-full max-ml:border-t max-ml:border-t-gray ml:w-[50%]">
                <TabList>
                  <SliderWrapper sliderSettings={sliderSettings}>
                    {props.fields.children.map((tab: HeroProductTabsProps, index: number) => {
                      return (
                        <TabHeading
                          key={index}
                          index={index}
                          onClick={() => {
                            setTabIndex(index);
                            window.location.hash =
                              tab.fields?.contentId?.value !== ''
                                ? tab.fields?.contentId?.value
                                : tab.id;
                          }}
                        >
                          <Text
                            tag={'h3'}
                            className={classNames(
                              themeData.classes.tabs.headlineClass,
                              index === tabIndex ? themeData.classes.tabs.selectedHeadlineClass : ''
                            )}
                            field={tab.fields.title}
                          />
                          {numberOfTabs > 3 &&
                            ((index === numberOfTabs - 1 && tabIndex !== 3) ||
                              (index === 0 && tabIndex !== 0)) && (
                              <div
                                className="absolute top-0 block h-full w-full content-[''] md:hidden"
                                style={{
                                  background:
                                    'linear-gradient(270deg, #ffffff 0%, rgba(255, 255, 255, 0) 100%)',
                                }}
                              ></div>
                            )}
                        </TabHeading>
                      );
                    })}
                  </SliderWrapper>
                </TabList>
              </div>
            </div>
          </div>
          {props.fields.children.map((tab: HeroProductTabsProps, index: number) => {
            return (
              <TabPanel forceRender={true} key={index}>
                <Placeholder
                  name="components"
                  rendering={tab}
                  render={(tabs) =>
                    tabs.map((tab, index) => (
                      <div key={index} className={classNames('col-span-12')}>
                        {tab}
                      </div>
                    ))
                  }
                  favoriteProducts={props.favoriteProducts}
                />
              </TabPanel>
            );
          })}
        </Tabs>
      </div>
    </Component>
  );
};

export default withDatasourceCheck()<HeroProductTabsProps>(HeroProductTabs);
