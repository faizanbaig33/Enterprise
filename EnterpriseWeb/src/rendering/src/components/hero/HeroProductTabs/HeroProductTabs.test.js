// Global
import { hasDataComponent, snapshot } from 'lib/jest/test-utils';
import 'lib/jest/match-media-mock';
// Local
import HeroProductTabs from './HeroProductTabs';
import defaultData from './HeroProductTabs.mock-data';

it('renders correctly', () => {
  const component = snapshot(HeroProductTabs, { componentProps: defaultData });
  hasDataComponent(component, 'hero/heroproducttabs');
});
