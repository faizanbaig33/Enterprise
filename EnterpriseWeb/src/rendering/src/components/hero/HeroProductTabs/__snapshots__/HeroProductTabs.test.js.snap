// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<section
  data-component="hero/heroproducttabs"
  id="id1695242607"
>
  <div
    class="my-8 w-full theme-white"
  >
    <div
      class="section-grid grid grid-cols-2 md:grid-cols-12 0 gap-s md:gap-s"
    >
      <div
        class=" col-span-12"
      >
        <div
          class="react-tabs"
          data-rttabs="true"
        >
          <div
            class="heroproducttabs-wrapper z-[70] w-full border-b border-b-gray bg-white"
          >
            <div
              class="flex w-full flex-col items-center justify-between ml:flex-row ml:px-m lg:mx-auto lg:max-w-screen-lg"
            >
              <div
                class="w-full ml:w-[50%]"
              >
                <h1
                  class="text-sm-s lg:text-s font-heavy font-sans uppercase max-ml:pl-ml max-ml:pb-xxs"
                >
                  Hero Product Tabs Headline
                </h1>
              </div>
              <div
                class="ml-auto w-full max-ml:border-t max-ml:border-t-gray ml:w-[50%]"
              >
                <ul
                  class="react-tabs__tab-list"
                  role="tablist"
                >
                  <div
                    class="slick-slider heroproducttabs slick-initialized"
                    dir="ltr"
                  >
                    <div
                      class="slick-list"
                    >
                      <div
                        class="slick-track"
                        style="opacity: 1; transform: translate3d(0px, 0px, 0px);"
                      >
                        <div
                          aria-hidden="false"
                          class="slick-slide slick-active slick-current"
                          data-index="0"
                          style="outline: none;"
                          tabindex="-1"
                        >
                          <div>
                            <div
                              class="react-tabs__tab flex h-full ml:min-w-[120px] relative justify-center text-center items-center px-xxs py-xs md:px-s border-b-8 w-[12rem] react-tabs__tab--selected border-b-primary "
                            >
                              <h3
                                class="font-heavy font-sans uppercase text-sm-xs ml:text-sm-s text-dark-gray hover:text-black text-base lg:text-sm-s justify-center text-center text-black"
                              >
                                Hero Product Tab Item 1
                              </h3>
                            </div>
                          </div>
                        </div>
                        <div
                          aria-hidden="false"
                          class="slick-slide slick-active"
                          data-index="1"
                          style="outline: none;"
                          tabindex="-1"
                        >
                          <div>
                            <div
                              class="react-tabs__tab flex h-full ml:min-w-[120px] relative justify-center text-center items-center px-xxs py-xs md:px-s border-b-8 w-[12rem] text-dark-gray border-b-white  "
                            >
                              <h3
                                class="font-heavy font-sans uppercase text-sm-xs ml:text-sm-s text-dark-gray hover:text-black text-base lg:text-sm-s justify-center text-center"
                              >
                                Hero Product Tab Item 2
                              </h3>
                            </div>
                          </div>
                        </div>
                        <div
                          aria-hidden="true"
                          class="slick-slide"
                          data-index="2"
                          style="outline: none;"
                          tabindex="-1"
                        >
                          <div>
                            <div
                              class="react-tabs__tab flex h-full ml:min-w-[120px] relative justify-center text-center items-center px-xxs py-xs md:px-s border-b-8 w-[12rem] text-dark-gray border-b-white  "
                            >
                              <h3
                                class="font-heavy font-sans uppercase text-sm-xs ml:text-sm-s text-dark-gray hover:text-black text-base lg:text-sm-s justify-center text-center"
                              >
                                Hero Product Tab Item 3
                              </h3>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </ul>
              </div>
            </div>
          </div>
          <div
            aria-labelledby="react-tabs-0"
            class="react-tabs__tab-panel react-tabs__tab-panel--selected"
            id="react-tabs-1"
            role="tabpanel"
          >
            <div
              class="col-span-12"
            >
              <div
                style="background: darkorange; outline: 5px solid orange; padding: 10px; color: white; max-width: 500px;"
              >
                <h2>
                  PromoGeneric
                </h2>
                <p>
                  JSS component is missing React implementation. See the developer console for more information.
                </p>
              </div>
            </div>
          </div>
          <div
            aria-labelledby="react-tabs-2"
            class="react-tabs__tab-panel"
            id="react-tabs-3"
            role="tabpanel"
          >
            <div
              class="col-span-12"
            >
              <div
                style="background: darkorange; outline: 5px solid orange; padding: 10px; color: white; max-width: 500px;"
              >
                <h2>
                  PromoGeneric
                </h2>
                <p>
                  JSS component is missing React implementation. See the developer console for more information.
                </p>
              </div>
            </div>
          </div>
          <div
            aria-labelledby="react-tabs-4"
            class="react-tabs__tab-panel"
            id="react-tabs-5"
            role="tabpanel"
          >
            <div
              class="col-span-12"
            >
              <div
                style="background: darkorange; outline: 5px solid orange; padding: 10px; color: white; max-width: 500px;"
              >
                <h2>
                  PromoGeneric
                </h2>
                <p>
                  JSS component is missing React implementation. See the developer console for more information.
                </p>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</section>
`;
